File: python_files/ex9.py
from abc import ABC, abstractmethod

# Classe abstrata Impressao
class Impressao(ABC):

    @abstractmethod
    def imprimir(self):
        pass

class Relatorio(Impressao):
    def __init__(self, titulo, conteudo, autor):
        self.titulo = titulo
        self.conteudo = conteudo
        self.autor = autor

    def imprimir(self):
        print("=== Relatório ===")
        print(f"Título: {self.titulo}")
        print(f"Conteúdo: {self.conteudo}")
        print(f"Autor: {self.autor}")
        print("=================")

class Contrato(Impressao):
    def __init__(self, partes, objeto, data_assinatura):
        self.partes = partes
        self.objeto = objeto
        self.data_assinatura = data_assinatura

    def imprimir(self):
        print("=== Contrato ===")
        print(f"Partes envolvidas: {self.partes}")
        print(f"Objeto do Contrato: {self.objeto}")
        print(f"Data de Assinatura: {self.data_assinatura}")
        print("================")

# Teste
relatorio = Relatorio("Relatório de Desempenho Anual", 
            "Este relatório contém uma análise detalhada dos KPIs e recomendações estratégicas para o próximo ano.", 
            "Carlos Souza")
contrato = Contrato("Empresa ABC e Empresa XYZ", 
            "Prestação de serviços de consultoria em tecnologia", 
            "15/09/2024")

relatorio.imprimir()
contrato.imprimir()

File: go_files/ex9.go
package main

import "fmt"

// Definindo a interface Impressao
type Impressao interface {
	Imprimir()
}

type Relatorio struct {
	Titulo   string
	Conteudo string
	Autor    string
}

func (r Relatorio) Imprimir() {
	fmt.Println("=== Relatório ===")
	fmt.Printf("Título: %s\n", r.Titulo)
	fmt.Printf("Conteúdo: %s\n", r.Conteudo)
	fmt.Printf("Autor: %s\n", r.Autor)
	fmt.Println("=================")
}

// Struct Contrato que implementa a interface Impressao
type Contrato struct {
	Partes          string
	Objeto          string
	DataAssinatura  string
}

func (c Contrato) Imprimir() {
	fmt.Println("=== Contrato ===")
	fmt.Printf("Partes envolvidas: %s\n", c.Partes)
	fmt.Printf("Objeto do Contrato: %s\n", c.Objeto)
	fmt.Printf("Data de Assinatura: %s\n", c.DataAssinatura)
	fmt.Println("================")
}

func main() {
	var impressao Impressao

	impressao = Relatorio{
		Titulo:   "Relatório de Desempenho Anual",
		Conteudo: "Este relatório contém uma análise detalhada dos KPIs e recomendações estratégicas para o próximo ano.",
		Autor:    "Carlos Souza",
	}
	impressao.Imprimir()

	impressao = Contrato{
		Partes:         "Empresa ABC e Empresa XYZ",
		Objeto:         "Prestação de serviços de consultoria em tecnologia",
		DataAssinatura: "15/09/2024",
	}
	impressao.Imprimir()
}